sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/display/DisplayData.java

```diff
@@ -508,55 +258,67 @@ public String getLabel() {
     @JsonGetter("linkUrl")
     @JsonInclude(JsonInclude.Include.NON_NULL)
     @Nullable
-    public String getLinkUrl() {
-      return url;
-    }
+    public abstract String getLinkUrl();
 
-    @Override
-    public String toString() {
-      return String.format("%s:%s=%s", ns, key, value);
+    public static <T> Item<T> create(String key, Type type, @Nullable T value) {
+      FormattedItemValue formatted = safeFormat(type, value);
+      return of(null, key, type, formatted.getLongValue(), formatted.getShortValue(), null, null);
     }
 
-    @Override
-    public boolean equals(Object obj) {
-      if (obj instanceof Item) {
-        Item that = (Item) obj;
-        return Objects.equals(this.ns, that.ns)
-            && Objects.equals(this.key, that.key)
-            && Objects.equals(this.type, that.type)
-            && Objects.equals(this.value, that.value)
-            && Objects.equals(this.shortValue, that.shortValue)
-            && Objects.equals(this.label, that.label)
-            && Objects.equals(this.url, that.url);
+    private static FormattedItemValue safeFormat(Type type, @Nullable Object value) {
+      if (value == null) {
+        return FormattedItemValue.DEFAULT;
       }
 
-      return false;
+      return type.format(value);
     }
 
-    @Override
-    public int hashCode() {
-      return Objects.hash(
-          this.ns,
-          this.key,
-          this.type,
-          this.value,
-          this.shortValue,
-          this.label,
-          this.url);
+    public Item<T> withNamespace(Class<?> namespace) {
+      checkNotNull(namespace);
+      return withNamespace(ClassForDisplay.of(namespace));
     }
 
-    private Item withLabel(String label) {
-      return new Item(this.ns, this.key, this.type, this.value, this.shortValue, this.url, label);
+    private Item<T> withNamespace(ClassForDisplay namespace) {
+      checkNotNull(namespace);
+      return withNamespace(namespaceOf(namespace));
```

Comment:
I think this enables removal of `ClassForDisplay`, right? The places it was used can now just create the appropriate set of `Item<T>`, and hold on to those, since we'll have done the namespace to String conversion up front?